File: impl/python-claudecode-sonnet4-1757409766/parser.py
#!/usr/bin/env python3
"""
Elf-lang parser implementation.
"""

import json
from typing import List, Dict, Any, Optional, Union
from lexer import Lexer, Token, TokenType


class Parser:
    def __init__(self, source: str):
        self.lexer = Lexer(source)
        self.tokens = self.lexer.tokenize()
        self.position = 0
        self.current_token = self.tokens[0] if self.tokens else None
    
    def advance(self):
        """Move to the next token."""
        self.position += 1
        if self.position >= len(self.tokens):
            self.current_token = None
        else:
            self.current_token = self.tokens[self.position]
    
    def peek(self, offset: int = 1) -> Optional[Token]:
        """Peek at the token at current position + offset."""
        peek_pos = self.position + offset
        if peek_pos >= len(self.tokens):
            return None
        return self.tokens[peek_pos]
    
    def expect(self, expected_type: TokenType) -> Token:
        """Expect a specific token type and consume it."""
        if self.current_token is None or self.current_token.type != expected_type:
            raise SyntaxError(f"Expected {expected_type}, got {self.current_token}")
        token = self.current_token
        self.advance()
        return token
    
    def skip_comments(self):
        """Skip any comment tokens."""
        while self.current_token and self.current_token.type == TokenType.CMT:
            self.advance()
    
    def parse(self) -> Dict[str, Any]:
        """Parse the program."""
        statements = []
        
        while self.current_token is not None:
            stmt = self.parse_statement()
            if stmt:
                statements.append(stmt)
        
        return {
            "statements": statements,
            "type": "Program"
        }
    
    def parse_statement(self) -> Optional[Dict[str, Any]]:
        """Parse a statement."""
        if self.current_token is None:
            return None
        
        # Handle comments
        if self.current_token.type == TokenType.CMT:
            comment_value = self.current_token.value
            self.advance()
            return {
                "type": "Comment",
                "value": comment_value
            }
        
        # Check for assignment: ID = expr
        if (self.current_token.type == TokenType.ID and 
            self.peek() and self.peek().type == TokenType.ASSIGN):
            name = {
                "name": self.current_token.value,
                "type": "Identifier"
            }
            self.advance()  # consume ID
            self.expect(TokenType.ASSIGN)  # consume =
            value = self.parse_expression()
            
            # Optional semicolon
            if self.current_token and self.current_token.type == TokenType.SEMICOLON:
                self.advance()
            
            return {
                "type": "Expression",
                "value": {
                    "type": "Assignment",
                    "name": name,
                    "value": value
                }
            }
        
        # Regular expression statement
        expr = self.parse_expression()
        
        # Optional semicolon
        if self.current_token and self.current_token.type == TokenType.SEMICOLON:
            self.advance()
        
        return {
            "type": "Expression",
            "value": expr
        }
    
    def parse_expression(self) -> Dict[str, Any]:
        """Parse an expression with precedence."""
        return self.parse_threading()
    
    def parse_threading(self) -> Dict[str, Any]:
        """Parse threading (|>) expressions (left-associative)."""
        left = self.parse_composition()
        
        if self.current_token and self.current_token.type == TokenType.PIPE_OP:
            # Collect all threading functions
            functions = []
            initial = left
            
            while self.current_token and self.current_token.type == TokenType.PIPE_OP:
                self.advance()  # consume |>
                right = self.parse_composition()
                functions.append(right)
            
            return {
                "functions": functions,
                "initial": initial,
                "type": "FunctionThread"
            }
        
        return left
    
    def parse_composition(self) -> Dict[str, Any]:
        """Parse composition (>>) expressions (right-associative)."""
        left = self.parse_logical_or()
        
        if self.current_token and self.current_token.type == TokenType.COMPOSE:
            # Collect all composition functions
            functions = [left]
            
            while self.current_token and self.current_token.type == TokenType.COMPOSE:
                self.advance()  # consume >>
                right = self.parse_logical_or()
                functions.append(right)
            
            return {
                "functions": functions,
                "type": "FunctionComposition"
            }
        
        return left
    
    def parse_logical_or(self) -> Dict[str, Any]:
        """Parse logical OR (||) expressions."""
        left = self.parse_logical_and()
        
        while self.current_token and self.current_token.type == TokenType.OR:
            operator = self.current_token.value
            self.advance()
            right = self.parse_logical_and()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_logical_and(self) -> Dict[str, Any]:
        """Parse logical AND (&&) expressions."""
        left = self.parse_equality()
        
        while self.current_token and self.current_token.type == TokenType.AND:
            operator = self.current_token.value
            self.advance()
            right = self.parse_equality()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_equality(self) -> Dict[str, Any]:
        """Parse equality expressions (==, !=)."""
        left = self.parse_comparison()
        
        while self.current_token and self.current_token.type in [TokenType.EQ, TokenType.NE]:
            operator = self.current_token.value
            self.advance()
            right = self.parse_comparison()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_comparison(self) -> Dict[str, Any]:
        """Parse comparison expressions (>, <, >=, <=)."""
        left = self.parse_addition()
        
        while self.current_token and self.current_token.type in [TokenType.GT, TokenType.LT, TokenType.GE, TokenType.LE]:
            operator = self.current_token.value
            self.advance()
            right = self.parse_addition()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_addition(self) -> Dict[str, Any]:
        """Parse addition and subtraction expressions."""
        left = self.parse_multiplication()
        
        while self.current_token and self.current_token.type in [TokenType.PLUS, TokenType.MINUS]:
            operator = self.current_token.value
            self.advance()
            right = self.parse_multiplication()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_multiplication(self) -> Dict[str, Any]:
        """Parse multiplication and division expressions."""
        left = self.parse_unary()
        
        while self.current_token and self.current_token.type in [TokenType.MULTIPLY, TokenType.DIVIDE]:
            operator = self.current_token.value
            self.advance()
            right = self.parse_unary()
            left = {
                "left": left,
                "operator": operator,
                "right": right,
                "type": "Infix"
            }
        
        return left
    
    def parse_unary(self) -> Dict[str, Any]:
        """Parse unary expressions."""
        if self.current_token and self.current_token.type == TokenType.MINUS:
            operator = self.current_token.value
            self.advance()
            operand = self.parse_unary()
            return {
                "type": "Prefix",
                "operator": operator,
                "operand": operand
            }
        
        return self.parse_postfix()
    
    def parse_postfix(self) -> Dict[str, Any]:
        """Parse postfix expressions (indexing, function calls)."""
        left = self.parse_primary()
        
        while True:
            if self.current_token and self.current_token.type == TokenType.LBRACKET:
                # Indexing
                self.advance()  # consume '['
                index = self.parse_expression()
                self.expect(TokenType.RBRACKET)
                left = {
                    "index": index,
                    "left": left,
                    "type": "Index"
                }
            elif self.current_token and self.current_token.type == TokenType.LPAREN:
                # Function call
                self.advance()  # consume '('
                arguments = []
                
                while self.current_token and self.current_token.type != TokenType.RPAREN:
                    arguments.append(self.parse_expression())
                    
                    if self.current_token and self.current_token.type == TokenType.COMMA:
                        self.advance()
                    elif self.current_token and self.current_token.type != TokenType.RPAREN:
                        break
                
                self.expect(TokenType.RPAREN)
                left = {
                    "arguments": arguments,
                    "function": left,
                    "type": "Call"
                }
            else:
                break
        
        return left
    
    def parse_primary(self) -> Dict[str, Any]:
        """Parse primary expressions."""
        if not self.current_token:
            raise SyntaxError("Unexpected end of input")
        
        # Literals
        if self.current_token.type == TokenType.INT:
            value = self.current_token.value
            self.advance()
            return {"type": "Integer", "value": value}
        
        elif self.current_token.type == TokenType.DEC:
            value = self.current_token.value
            self.advance()
            return {"type": "Decimal", "value": value}
        
        elif self.current_token.type == TokenType.STR:
            # Remove quotes and handle escape sequences for the value
            raw_value = self.current_token.value
            if raw_value.startswith('"') and raw_value.endswith('"'):
                # Extract content without quotes
                content = raw_value[1:-1]
                # Handle escape sequences
                unescaped = content.replace('\\\"', '"').replace('\\\\', '\\').replace('\\n', '\n').replace('\\t', '\t')
                self.advance()
                return {"type": "String", "value": unescaped}
            else:
                self.advance()
                return {"type": "String", "value": raw_value}
        
        elif self.current_token.type == TokenType.TRUE:
            self.advance()
            return {"type": "Boolean", "value": True}
        
        elif self.current_token.type == TokenType.FALSE:
            self.advance()
            return {"type": "Boolean", "value": False}
        
        elif self.current_token.type == TokenType.NIL:
            self.advance()
            return {"type": "Nil"}
        
        # Identifiers
        elif self.current_token.type == TokenType.ID:
            name = self.current_token.value
            self.advance()
            return {"name": name, "type": "Identifier"}
        
        # Parentheses
        elif self.current_token.type == TokenType.LPAREN:
            self.advance()
            expr = self.parse_expression()
            self.expect(TokenType.RPAREN)
            return expr
        
        # Let declarations
        elif self.current_token.type == TokenType.LET:
            return self.parse_let()
        
        # If expressions
        elif self.current_token.type == TokenType.IF:
            return self.parse_if()
        
        # Lists
        elif self.current_token.type == TokenType.LBRACKET:
            return self.parse_list()
        
        # Sets and Dictionaries
        elif self.current_token.type == TokenType.LBRACE:
            return self.parse_set()
        
        elif self.current_token.type == TokenType.DICT_START:  # #{
            return self.parse_dictionary()
        
        # Function literals
        elif self.current_token.type == TokenType.PIPE:
            return self.parse_function()
        
        # Zero-parameter function literals (||)
        elif self.current_token.type == TokenType.OR:
            return self.parse_zero_param_function()
        
        # Operator identifiers (for higher-order functions)
        elif self.current_token.type in [TokenType.PLUS, TokenType.MINUS, TokenType.MULTIPLY, TokenType.DIVIDE]:
            name = self.current_token.value
            self.advance()
            return {"name": name, "type": "Identifier"}
        
        else:
            raise SyntaxError(f"Unexpected token: {self.current_token}")
    
    def parse_let(self) -> Dict[str, Any]:
        """Parse let declarations."""
        self.expect(TokenType.LET)
        
        is_mutable = False
        if self.current_token and self.current_token.type == TokenType.MUT:
            is_mutable = True
            self.advance()
        
        if not self.current_token or self.current_token.type != TokenType.ID:
            raise SyntaxError(f"Expected identifier after let, got {self.current_token}")
        
        name = {
            "name": self.current_token.value,
            "type": "Identifier"
        }
        self.advance()
        
        self.expect(TokenType.ASSIGN)
        value = self.parse_expression()
        
        return {
            "name": name,
            "type": "MutableLet" if is_mutable else "Let",
            "value": value
        }
    
    def parse_if(self) -> Dict[str, Any]:
        """Parse if expressions."""
        self.expect(TokenType.IF)
        condition = self.parse_expression()
        consequence = self.parse_block()
        
        alternative = None
        if self.current_token and self.current_token.type == TokenType.ELSE:
            self.advance()
            alternative = self.parse_block()
        
        return {
            "alternative": alternative,
            "condition": condition,
            "consequence": consequence,
            "type": "If"
        }
    
    def parse_block(self) -> Dict[str, Any]:
        """Parse a block expression."""
        self.expect(TokenType.LBRACE)
        statements = []
        
        while self.current_token and self.current_token.type != TokenType.RBRACE:
            self.skip_comments()
            if self.current_token and self.current_token.type != TokenType.RBRACE:
                stmt = self.parse_statement()
                if stmt:
                    statements.append(stmt)
        
        self.expect(TokenType.RBRACE)
        
        return {
            "statements": statements,
            "type": "Block"
        }
    
    def parse_list(self) -> Dict[str, Any]:
        """Parse list literals."""
        self.expect(TokenType.LBRACKET)
        items = []
        
        while self.current_token and self.current_token.type != TokenType.RBRACKET:
            items.append(self.parse_expression())
            
            if self.current_token and self.current_token.type == TokenType.COMMA:
                self.advance()
            elif self.current_token and self.current_token.type != TokenType.RBRACKET:
                break
        
        self.expect(TokenType.RBRACKET)
        
        return {
            "items": items,
            "type": "List"
        }
    
    def parse_set(self) -> Dict[str, Any]:
        """Parse set literals."""
        self.expect(TokenType.LBRACE)
        items = []
        
        while self.current_token and self.current_token.type != TokenType.RBRACE:
            items.append(self.parse_expression())
            
            if self.current_token and self.current_token.type == TokenType.COMMA:
                self.advance()
            elif self.current_token and self.current_token.type != TokenType.RBRACE:
                break
        
        self.expect(TokenType.RBRACE)
        
        return {
            "items": items,
            "type": "Set"
        }
    
    def parse_dictionary(self) -> Dict[str, Any]:
        """Parse dictionary literals."""
        self.expect(TokenType.DICT_START)  # #{
        items = []
        
        while self.current_token and self.current_token.type != TokenType.RBRACE:
            key = self.parse_expression()
            self.expect(TokenType.COLON)
            value = self.parse_expression()
            
            items.append({
                "key": key,
                "value": value
            })
            
            if self.current_token and self.current_token.type == TokenType.COMMA:
                self.advance()
            elif self.current_token and self.current_token.type != TokenType.RBRACE:
                break
        
        self.expect(TokenType.RBRACE)
        
        return {
            "items": items,
            "type": "Dictionary"
        }
    
    def parse_function(self) -> Dict[str, Any]:
        """Parse function literals."""
        self.expect(TokenType.PIPE)
        parameters = []
        
        # Parse parameters
        while self.current_token and self.current_token.type != TokenType.PIPE:
            if self.current_token.type == TokenType.ID:
                parameters.append({
                    "name": self.current_token.value,
                    "type": "Identifier"
                })
                self.advance()
                
                if self.current_token and self.current_token.type == TokenType.COMMA:
                    self.advance()
            else:
                break
        
        self.expect(TokenType.PIPE)
        
        # Parse body - could be expression or block
        if self.current_token and self.current_token.type == TokenType.LBRACE:
            body = self.parse_block()
        else:
            # Single expression body - wrap in a block
            expr = self.parse_expression()
            body = {
                "statements": [{"type": "Expression", "value": expr}],
                "type": "Block"
            }
        
        return {
            "body": body,
            "parameters": parameters,
            "type": "Function"
        }
    
    def parse_zero_param_function(self) -> Dict[str, Any]:
        """Parse zero-parameter function literals (||)."""
        self.expect(TokenType.OR)  # consume ||
        
        # Parse body - could be expression or block
        if self.current_token and self.current_token.type == TokenType.LBRACE:
            body = self.parse_block()
        else:
            # Single expression body - wrap in a block
            expr = self.parse_expression()
            body = {
                "statements": [{"type": "Expression", "value": expr}],
                "type": "Block"
            }
        
        return {
            "body": body,
            "parameters": [],  # Zero parameters
            "type": "Function"
        }


def parse_to_json(source: str) -> str:
    """Parse source and return pretty-printed JSON."""
    parser = Parser(source)
    ast = parser.parse()
    
    return json.dumps(ast, indent=2, separators=(',', ': '), ensure_ascii=False)

File: impl/scala-claudecode-sonnet4-1757836384/src/main/scala/elflang/Parser.scala
package elflang

import scala.collection.mutable.ListBuffer

class ParseError(message: String) extends Exception(message)

object Parser:
  def parse(tokens: List[Token]): ujson.Value =
    val parser = new Parser(tokens)
    parser.parseProgram()

class Parser(tokens: List[Token]):
  private var pos = 0
  private val length = tokens.length

  private def current: Option[Token] =
    if pos < length then Some(tokens(pos)) else None

  private def peek(offset: Int = 1): Option[Token] =
    val newPos = pos + offset
    if newPos < length then Some(tokens(newPos)) else None

  private def advance(): Option[Token] =
    val token = current
    pos += 1
    token

  private def expect(tokenType: String): Token =
    current match
      case Some(token) if token.tokenType == tokenType =>
        advance()
        token
      case Some(token) =>
        throw ParseError(s"Expected $tokenType, got ${token.tokenType}")
      case None =>
        throw ParseError(s"Expected $tokenType, got end of input")

  private def match_(tokenTypes: String*): Boolean =
    current.exists(token => tokenTypes.contains(token.tokenType))

  def parseProgram(): ujson.Value =
    val statements = ListBuffer[ujson.Value]()

    while pos < length do
      // Include comments as statements
      if match_("CMT") then
        val comment = advance().get
        statements += ujson.Obj(
          "type" -> "Comment",
          "value" -> comment.value
        )
      else
        statements += parseStatement()

    ujson.Obj(
      "statements" -> ujson.Arr(statements.toSeq*),
      "type" -> "Program"
    )

  private def parseStatement(): ujson.Value =
    // Check for assignment statements (mutable variable updates)
    if match_("ID") then
      val peekNext = peek()
      peekNext match
        case Some(Token("=", _)) =>
          val name = advance().get
          advance() // consume '='
          val value = parseExpression()
          // Optional semicolon
          if match_(";") then
            advance()
          return ujson.Obj(
            "type" -> "Expression",
            "value" -> ujson.Obj(
              "name" -> ujson.Obj("name" -> name.value, "type" -> "Identifier"),
              "type" -> "Assignment",
              "value" -> value
            )
          )
        case _ => // Continue with normal expression

    val expr = parseExpression()

    // Optional semicolon
    if match_(";") then
      advance()

    ujson.Obj(
      "type" -> "Expression",
      "value" -> expr
    )

  private def parseExpression(): ujson.Value =
    parseFunctionThread()

  // Precedence levels (lowest to highest):
  // Function thread: |>
  // Function composition: >>
  // Logical OR: ||
  // Logical AND: &&
  // Comparison: ==, !=, >, <, >=, <=
  // Addition/Subtraction: + -
  // Multiplication/Division: * /
  // Primary: literals, parentheses, identifiers, let expressions

  private def parseFunctionThread(): ujson.Value =
    var left = parseFunctionComposition()

    while match_("|>") do
      advance() // consume '|>'
      val right = parseFunctionComposition()
      // Build up the thread
      left match
        case thread @ ujson.Obj(pairs) if pairs.value.get("type").contains(ujson.Str("FunctionThread")) =>
          // Already a thread, add to it
          val functions = thread("functions").arr.toSeq
          left = ujson.Obj(
            "functions" -> ujson.Arr((functions :+ right)*),
            "initial" -> thread("initial"),
            "type" -> "FunctionThread"
          )
        case _ =>
          // Start a new thread
          left = ujson.Obj(
            "functions" -> ujson.Arr(right),
            "initial" -> left,
            "type" -> "FunctionThread"
          )

    left

  private def parseFunctionComposition(): ujson.Value =
    var functions = ListBuffer[ujson.Value]()
    functions += parseLogicalOr()

    while match_(">>") do
      advance() // consume '>>'
      functions += parseLogicalOr()

    if functions.length > 1 then
      ujson.Obj(
        "functions" -> ujson.Arr(functions.toSeq*),
        "type" -> "FunctionComposition"
      )
    else
      functions.head

  private def parseLogicalOr(): ujson.Value =
    var left = parseLogicalAnd()

    while match_("||") do
      val op = advance().get.value
      val right = parseLogicalAnd()
      left = ujson.Obj(
        "left" -> left,
        "operator" -> op,
        "right" -> right,
        "type" -> "Infix"
      )

    left

  private def parseLogicalAnd(): ujson.Value =
    var left = parseComparison()

    while match_("&&") do
      val op = advance().get.value
      val right = parseComparison()
      left = ujson.Obj(
        "left" -> left,
        "operator" -> op,
        "right" -> right,
        "type" -> "Infix"
      )

    left

  private def parseComparison(): ujson.Value =
    var left = parseAddition()

    while match_("==", "!=", ">", "<", ">=", "<=") do
      val op = advance().get.value
      val right = parseAddition()
      left = ujson.Obj(
        "left" -> left,
        "operator" -> op,
        "right" -> right,
        "type" -> "Infix"
      )

    left

  private def parseAddition(): ujson.Value =
    var left = parseMultiplication()

    while match_("+", "-") do
      val op = advance().get.value
      val right = parseMultiplication()
      left = ujson.Obj(
        "left" -> left,
        "operator" -> op,
        "right" -> right,
        "type" -> "Infix"
      )

    left

  private def parseMultiplication(): ujson.Value =
    var left = parsePostfix()

    while match_("*", "/") do
      val op = advance().get.value
      val right = parsePostfix()
      left = ujson.Obj(
        "left" -> left,
        "operator" -> op,
        "right" -> right,
        "type" -> "Infix"
      )

    left

  private def parsePostfix(): ujson.Value =
    var left = parsePrimary()

    // Handle indexing and function calls (higher precedence)
    while match_("[", "(") do
      current match
        case Some(Token("[", _)) =>
          advance() // consume '['
          val index = parseExpression()
          expect("]")
          left = ujson.Obj(
            "index" -> index,
            "left" -> left,
            "type" -> "Index"
          )
        case Some(Token("(", _)) =>
          advance() // consume '('
          val arguments = ListBuffer[ujson.Value]()

          if !match_(")") then
            arguments += parseExpression()
            while match_(",") do
              advance() // consume ','
              if match_(")") then
                // Allow trailing comma
                ()
              else
                arguments += parseExpression()

          expect(")")
          left = ujson.Obj(
            "arguments" -> ujson.Arr(arguments.toSeq*),
            "function" -> left,
            "type" -> "Call"
          )
        case _ =>
          return left

    left

  private def parsePrimary(): ujson.Value =
    current match
      // Handle unary minus
      case Some(Token("-", _)) =>
        advance() // consume '-'
        val operand = parsePrimary()
        ujson.Obj(
          "operand" -> operand,
          "operator" -> "-",
          "type" -> "Prefix"
        )

      case Some(Token("INT", value)) =>
        advance()
        // Keep underscores in the AST representation
        ujson.Obj("type" -> "Integer", "value" -> value)

      case Some(Token("DEC", value)) =>
        advance()
        // Keep underscores in the AST representation
        ujson.Obj("type" -> "Decimal", "value" -> value)

      case Some(Token("STR", value)) =>
        advance()
        // Remove quotes and unescape
        val content = unescapeString(value.substring(1, value.length - 1))
        ujson.Obj("type" -> "String", "value" -> content)

      case Some(Token("TRUE", _)) =>
        advance()
        ujson.Obj("type" -> "Boolean", "value" -> true)

      case Some(Token("FALSE", _)) =>
        advance()
        ujson.Obj("type" -> "Boolean", "value" -> false)

      case Some(Token("NIL", _)) =>
        advance()
        ujson.Obj("type" -> "Nil")

      case Some(Token("ID", value)) =>
        advance()
        ujson.Obj("name" -> value, "type" -> "Identifier")

      // Handle operators as identifiers in certain contexts (but not unary minus)
      case Some(Token(op, value)) if "+*/<>=".contains(op) && op.length == 1 =>
        advance()
        ujson.Obj("name" -> value, "type" -> "Identifier")

      case Some(Token("LET", _)) =>
        parseLet()

      case Some(Token("IF", _)) =>
        parseIf()

      case Some(Token("(", _)) =>
        advance() // consume '('
        val expr = parseExpression()
        expect(")")
        expr

      case Some(Token("[", _)) =>
        parseList()

      case Some(Token("{", _)) =>
        parseSet()

      case Some(Token("#{", _)) =>
        parseDictionary()

      case Some(Token("|", _)) =>
        parseFunction()

      case Some(Token("||", _)) =>
        parseZeroArgFunction()

      case Some(token) =>
        throw ParseError(s"Unexpected token: ${token.tokenType}")

      case None =>
        throw ParseError("Unexpected end of input")

  private def parseLet(): ujson.Value =
    advance() // consume 'let'

    val isMutable = if match_("MUT") then
      advance()
      true
    else
      false

    val name = expect("ID")
    expect("=")
    val value = parseExpression()

    ujson.Obj(
      "name" -> ujson.Obj("name" -> name.value, "type" -> "Identifier"),
      "type" -> (if isMutable then "MutableLet" else "Let"),
      "value" -> value
    )

  private def parseList(): ujson.Value =
    advance() // consume '['
    val items = ListBuffer[ujson.Value]()

    if !match_("]") then
      items += parseExpression()
      while match_(",") do
        advance() // consume ','
        if match_("]") then
          // Allow trailing comma but don't add another element
          ()
        else
          items += parseExpression()

    expect("]")
    ujson.Obj(
      "items" -> ujson.Arr(items.toSeq*),
      "type" -> "List"
    )

  private def parseSet(): ujson.Value =
    advance() // consume '{'
    val items = ListBuffer[ujson.Value]()

    if !match_("}") then
      items += parseExpression()
      while match_(",") do
        advance() // consume ','
        if match_("}") then
          // Allow trailing comma but don't add another element
          ()
        else
          items += parseExpression()

    expect("}")
    ujson.Obj(
      "items" -> ujson.Arr(items.toSeq*),
      "type" -> "Set"
    )

  private def parseDictionary(): ujson.Value =
    advance() // consume '#{'
    val items = ListBuffer[ujson.Value]()

    if !match_("}") then
      // Parse key-value pair
      val key = parseExpression()
      expect(":")
      val value = parseExpression()
      items += ujson.Obj("key" -> key, "value" -> value)

      while match_(",") do
        advance() // consume ','
        if match_("}") then
          // Allow trailing comma but don't add another element
          ()
        else
          val nextKey = parseExpression()
          expect(":")
          val nextValue = parseExpression()
          items += ujson.Obj("key" -> nextKey, "value" -> nextValue)

    expect("}")
    ujson.Obj(
      "items" -> ujson.Arr(items.toSeq*),
      "type" -> "Dictionary"
    )

  private def parseIf(): ujson.Value =
    advance() // consume 'if'
    val condition = parseExpression()
    val consequence = parseBlock()
    expect("ELSE")
    val alternative = parseBlock()

    ujson.Obj(
      "alternative" -> alternative,
      "condition" -> condition,
      "consequence" -> consequence,
      "type" -> "If"
    )

  private def parseFunction(): ujson.Value =
    advance() // consume '|'
    val parameters = ListBuffer[ujson.Value]()

    // Parse parameter list
    if !match_("|") then
      parameters += ujson.Obj("name" -> expect("ID").value, "type" -> "Identifier")
      while match_(",") do
        advance() // consume ','
        parameters += ujson.Obj("name" -> expect("ID").value, "type" -> "Identifier")

    expect("|") // closing |

    // Parse body - either a block or a single expression wrapped in a block
    val body = if match_("{") then
      parseBlock()
    else
      // Single expression - wrap in a block
      val expr = parseExpression()
      ujson.Obj(
        "statements" -> ujson.Arr(ujson.Obj(
          "type" -> "Expression",
          "value" -> expr
        )),
        "type" -> "Block"
      )

    ujson.Obj(
      "body" -> body,
      "parameters" -> ujson.Arr(parameters.toSeq*),
      "type" -> "Function"
    )

  private def parseZeroArgFunction(): ujson.Value =
    advance() // consume '||'

    // Parse body - either a block or a single expression wrapped in a block
    val body = if match_("{") then
      parseBlock()
    else
      // Single expression - wrap in a block
      val expr = parseExpression()
      ujson.Obj(
        "statements" -> ujson.Arr(ujson.Obj(
          "type" -> "Expression",
          "value" -> expr
        )),
        "type" -> "Block"
      )

    ujson.Obj(
      "body" -> body,
      "parameters" -> ujson.Arr(), // Empty parameters for zero-arg function
      "type" -> "Function"
    )

  private def parseBlock(): ujson.Value =
    expect("{")
    val statements = ListBuffer[ujson.Value]()

    while !match_("}") do
      // Skip comments at block level
      if match_("CMT") then
        advance()
      else
        statements += parseBlockStatement()

    expect("}")
    ujson.Obj(
      "statements" -> ujson.Arr(statements.toSeq*),
      "type" -> "Block"
    )

  private def parseBlockStatement(): ujson.Value =
    // Check for assignment statements (mutable variable updates)
    if match_("ID") then
      val peekNext = peek()
      peekNext match
        case Some(Token("=", _)) =>
          val name = advance().get
          advance() // consume '='
          val value = parseExpression()
          // Optional semicolon
          if match_(";") then
            advance()
          return ujson.Obj(
            "type" -> "Expression",
            "value" -> ujson.Obj(
              "name" -> ujson.Obj("name" -> name.value, "type" -> "Identifier"),
              "type" -> "Assignment",
              "value" -> value
            )
          )
        case _ => // Continue with normal expression

    val expr = parseExpression()

    // Optional semicolon
    if match_(";") then
      advance()

    ujson.Obj(
      "type" -> "Expression",
      "value" -> expr
    )

  private def unescapeString(s: String): String =
    val sb = new StringBuilder()
    var i = 0
    while i < s.length do
      if s(i) == '\\' && i < s.length - 1 then
        s(i + 1) match
          case 'n' => sb.append('\n')
          case 't' => sb.append('\t')
          case '"' => sb.append('"')
          case '\\' => sb.append('\\')
          case c => sb.append('\\').append(c)
        i += 2
      else
        sb.append(s(i))
        i += 1
    sb.toString